exporter/opentelemetry-exporter-prometheus/src/opentelemetry/exporter/prometheus/_mapping.py

```diff
@@ -0,0 +1,135 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from re import UNICODE, compile
+
+_SANITIZE_NAME_RE = compile(r"([^a-zA-Z0-9:]+)|_{2,}", UNICODE)
+# Same as name, but doesn't allow ":"
+_SANITIZE_ATTRIBUTE_KEY_RE = compile(r"([^a-zA-Z0-9]+)|_{2,}", UNICODE)
+
+# UCUM annotations are ASCII chars 33-126 enclosed in curly braces
+# https://ucum.org/ucum#para-6
+_UCUM_ANNOTATION_CURLY = compile(r"{[!-~]*}")
+
+# Remaps common UCUM and SI units to prometheus conventions. Copied from
+# https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/v0.101.0/pkg/translator/prometheus/normalize_name.go#L19
+# See specification:
+# https://github.com/open-telemetry/opentelemetry-specification/blob/v1.33.0/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1
+_UNIT_MAPPINGS = {
+    # Time
+    "d": "days",
+    "h": "hours",
+    "min": "minutes",
+    "s": "seconds",
+    "ms": "milliseconds",
+    "us": "microseconds",
+    "ns": "nanoseconds",
+    # Bytes
+    "By": "bytes",
+    "KiBy": "kibibytes",
+    "MiBy": "mebibytes",
+    "GiBy": "gibibytes",
+    "TiBy": "tibibytes",
+    "KBy": "kilobytes",
+    "MBy": "megabytes",
+    "GBy": "gigabytes",
+    "TBy": "terabytes",
+    # SI
+    "m": "meters",
+    "V": "volts",
+    "A": "amperes",
+    "J": "joules",
+    "W": "watts",
+    "g": "grams",
+    # Misc
+    "Cel": "celsius",
+    "Hz": "hertz",
+    # TODO: this conflicts with the spec but I think it is correct. Need to open a spec issue
+    "1": "",
+    "%": "percent",
+}
+# Similar to _UNIT_MAPPINGS, but for "per" unit denominator.
+# Example: s => per second (singular)
+# Copied from https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/80317ce83ed87a2dff0c316bb939afbfaa823d5e/pkg/translator/prometheus/normalize_name.go#L58
+_PER_UNIT_MAPPINGS = {
+    "s": "second",
+    "m": "minute",
+    "h": "hour",
+    "d": "day",
+    "w": "week",
+    "mo": "month",
+    "y": "year",
+}
+
+
+def sanitize_full_name(name: str) -> str:
+    """sanitize the given metric name according to Prometheus rule, including sanitizing
+    leading digits
+
+    https://github.com/open-telemetry/opentelemetry-specification/blob/v1.33.0/specification/compatibility/prometheus_and_openmetrics.md#metric-metadata-1
+    """
+    # Leading number special case
+    if name and name[0].isdigit():
```

Comment:
From the specs:

> The metric name is required to match the regex: [a-zA-Z_:]([a-zA-Z0-9_:])*

So only the FIRST digit must not be a number.

Reply:
The regex is `[a-zA-Z_:]([a-zA-Z0-9_:])*`, see [here](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.33.0/specification/compatibility/prometheus_and_openmetrics.md#otlp-metric-points-to-prometheus:~:text=The%20Name%20of%20an%20OTLP%20metric%20MUST%20be%20added%20as%20the%20Prometheus%20Metric%20Name%2C%20with%20unit%20and%20type%20suffixes%20added%20as%20described%20below.%20The%20metric%20name%20is%20required%20to%20match%20the%20regex%3A%20%5Ba%2DzA%2DZ_%3A%5D(%5Ba%2DzA%2DZ0%2D9_%3A%5D)*).

I could probably try to cram this into the substitution regex but it would probably be hard to read

Reply:
> I could probably try to cram this into the substitution regex but it would probably be hard to read

I tried this out and it was hard to read and much slower. I simplified the regex a bit to remove the special `_` case